File: examples/complex.R
## Benchmark polytope for the R package vertexenum -- the exact counterpart of
## examples/complex.py in this repository.
##
## Builds the identical polytope (the box [-2, 2]^10 intersected with 28
## pseudo-random integer cutting planes), enumerates its vertices with the R
## package, verifies the result against the checksums produced by the Python
## implementation, and reports the elapsed time for benchmarking.
##
## Requirements: the archived CRAN package `vertexenum` (and its dependency
## `numbers`). Install one of these ways (Rtools needed on Windows):
##   install.packages("numbers")
##   install.packages("r_package", repos = NULL, type = "source")   # from repo root
## or from the CRAN archive:
##   install.packages("https://cran.r-project.org/src/contrib/Archive/vertexenum/vertexenum_1.0.2.tar.gz",
##                    repos = NULL, type = "source")
##
## Run:  Rscript examples/complex.R

suppressPackageStartupMessages(library(vertexenum))

DIM <- 10L        # number of variables
CUTS <- 28L       # number of pseudo-random cutting planes
BOX <- 2L         # box is [-BOX, BOX]^DIM
SEED <- 20260719

## Reference values from examples/complex.py (Python + lrslib v5.1):
EXPECTED_VERTICES <- 40317L
EXPECTED_COORD_SUM <- -8522.372864
EXPECTED_ABS_SUM <- 678450.682632
CHECKSUM_TOL <- 1e-3

## Deterministic coefficient stream, identical to complex.py:
##   x_{k+1} = (1103515245 * x_k + 12345) mod 2^31,  x_0 = SEED
##   value_k = (x_{k+1} mod 1999) - 999
## R has no 64-bit integers and 1103515245 * x overflows the 2^53 exact-double
## range, so the multiplication is split into 16-bit halves; every
## intermediate stays below 2^48 and the arithmetic remains exact.
lcg_ints <- function(count, seed = SEED) {
  MULT <- 1103515245
  INC <- 12345
  MOD <- 2^31
  x <- seed
  out <- integer(count)
  for (k in seq_len(count)) {
    hi <- x %/% 65536            # < 2^15
    lo <- x %% 65536             # < 2^16
    x <- ((MULT * hi) %% MOD * 65536 + MULT * lo + INC) %% MOD
    out[k] <- x %% 1999 - 999
  }
  out
}

## Same construction as complex.py: CUTS rows of LCG coefficients (row-major)
## with b_i = sum_j |a_ij|, then the box constraints +-x_j <= BOX.
build_polytope <- function(dim = DIM, cuts = CUTS, box = BOX) {
  vals <- lcg_ints(cuts * dim)
  m <- cuts + 2L * dim
  A <- matrix(0, nrow = m, ncol = dim)
  b <- numeric(m)
  for (i in seq_len(cuts)) {
    row <- vals[((i - 1L) * dim + 1L):(i * dim)]
    if (all(row == 0)) row[1L] <- 1
    A[i, ] <- row
    b[i] <- sum(abs(row))
  }
  r <- cuts
  for (j in seq_len(dim)) {
    for (s in c(1, -1)) {
      r <- r + 1L
      A[r, j] <- s
      b[r] <- box
    }
  }
  list(A = A, b = b)
}

check <- function(label, ok, detail) {
  cat(sprintf("%-14s %s  (%s)\n", label, if (ok) "OK" else "FAIL", detail))
  ok
}

p <- build_polytope()
cat(sprintf("R %s, vertexenum %s, polytope: dim=%d, %d inequalities (%d cuts + %d box)\n",
            getRversion(), as.character(packageVersion("vertexenum")),
            DIM, nrow(p$A), CUTS, 2L * DIM))

timing <- system.time(V <- enumerate.vertices(p$A, p$b))
elapsed <- timing[["elapsed"]]

n_vertices <- nrow(V)
coord_sum <- sum(V)
abs_sum <- sum(abs(V))

cat(sprintf("%d vertices in %.2f s\n", n_vertices, elapsed))
cat(sprintf("checksum: coordinate sum = %.6f, abs sum = %.6f\n",
            coord_sum, abs_sum))

all_ok <- all(
  check("vertex count", n_vertices == EXPECTED_VERTICES,
        sprintf("%d vs %d", n_vertices, EXPECTED_VERTICES)),
  check("coord sum", abs(coord_sum - EXPECTED_COORD_SUM) < CHECKSUM_TOL,
        sprintf("%.6f vs %.6f", coord_sum, EXPECTED_COORD_SUM)),
  check("abs sum", abs(abs_sum - EXPECTED_ABS_SUM) < CHECKSUM_TOL,
        sprintf("%.6f vs %.6f", abs_sum, EXPECTED_ABS_SUM))
)

cat(sprintf("benchmark: R/vertexenum elapsed = %.2f s (Python/lrslib v5.1 reference: ~8 s)\n",
            elapsed))

if (!all_ok) {
  cat("checksum mismatch: R result differs from the Python reference\n")
  quit(status = 1L)
}

File: examples/complex.py
"""Benchmark polytope: a randomly cut box in 10 dimensions.

The polytope is the box [-2, 2]^10 intersected with 28 pseudo-random
halfspaces a_i . x <= b_i, where the coefficients a_ij are integers in
[-999, 999] drawn from a small linear congruential generator and
b_i = sum_j |a_ij| (the cutting plane sits halfway between the origin and
the deepest box corner, so every cut slices the box, creating many
vertices). On the reference machine (2026 laptop, Windows, lrslib v5.1
with lrsmp arithmetic) enumeration takes roughly 8 seconds and yields
40317 vertices.

Everything is deterministic and integer-valued, so the exact same
polytope can be rebuilt in other environments for benchmarking — e.g. in
the R package vertexenum, or against a newer lrslib. To port the
generator, only the LCG below is needed:

    x_{k+1} = (1103515245 * x_k + 12345) mod 2^31,  x_0 = 20260719
    value_k = (x_{k+1} mod 1999) - 999

(the classic glibc rand() constants; all arithmetic fits in 64-bit
integers). Row i of A is values dim*i .. dim*i+dim-1, read row-major.

Run with --ine FILE to also write the system in lrs .ine format, for
feeding directly to a standalone lrs binary.
"""

import argparse
import time

from vertexenum import enumerate_vertices, lrs_version

DIM = 10        # number of variables
CUTS = 28       # number of pseudo-random cutting planes
BOX = 2         # box is [-BOX, BOX]^DIM
SEED = 20260719

LCG_MULT = 1103515245
LCG_INC = 12345
LCG_MOD = 2**31


def lcg_ints(count, seed=SEED):
    """First `count` values of the deterministic coefficient stream."""
    x = seed
    out = []
    for _ in range(count):
        x = (LCG_MULT * x + LCG_INC) % LCG_MOD
        out.append(x % 1999 - 999)
    return out


def build_polytope(dim=DIM, cuts=CUTS, box=BOX):
    """Return (A, b) for the benchmark polytope, all entries integers."""
    vals = lcg_ints(cuts * dim)
    A = []
    b = []
    for i in range(cuts):
        row = vals[i * dim:(i + 1) * dim]
        if all(v == 0 for v in row):  # never happens for the default seed
            row[0] = 1
        A.append(row)
        b.append(sum(abs(v) for v in row))
    for j in range(dim):
        for sign in (1, -1):
            row = [0] * dim
            row[j] = sign
            A.append(row)
            b.append(box)
    return A, b


def write_ine(path, A, b):
    """Write the system in lrs .ine (H-representation) format."""
    with open(path, "w") as f:
        f.write("random-cut-polytope\n")
        f.write("H-representation\n")
        f.write("begin\n")
        f.write(f"{len(A)} {len(A[0]) + 1} rational\n")
        for row, rhs in zip(A, b):
            f.write(" ".join([str(rhs)] + [str(-v) for v in row]) + "\n")
        f.write("end\n")


def main():
    parser = argparse.ArgumentParser(description=__doc__.splitlines()[0])
    parser.add_argument("--ine", metavar="FILE",
                        help="also write the system in lrs .ine format")
    args = parser.parse_args()

    A, b = build_polytope()
    print(f"{lrs_version()}, polytope: dim={DIM}, {len(A)} inequalities "
          f"({CUTS} cuts + {2 * DIM} box)")
    if args.ine:
        write_ine(args.ine, A, b)
        print(f"wrote {args.ine}")

    t0 = time.perf_counter()
    vertices = enumerate_vertices(A, b)
    elapsed = time.perf_counter() - t0

    # order-independent figures for cross-implementation comparison
    print(f"{vertices.shape[0]} vertices in {elapsed:.2f} s")
    print(f"checksum: coordinate sum = {vertices.sum():.6f}, "
          f"abs sum = {abs(vertices).sum():.6f}")


if __name__ == "__main__":
    main()
